design: ConfigStore design (#2536)
NVIDIA/bare-metal-manager-core: docs/design/config-store-design/config-store-design.md

Example: `RouteServersConfig` implements `Removable` because removing a route server has no effect on other resources.

### `Reconcilable`

Extends `Additive`. Implemented by types whose seeded definitions can drift from their config file declarations over time. Replaces the current silent-warn-and-ignore behavior in `reconcile_pool_defs` and `reconcile_network_defs` with observable, operator-controlled reconciliation.

```rust
pub enum ReconcileAction {
    AutoApply,            // safe to apply on startup without operator intervention
    RequireConfirmation,  // surface the drift visibly; don't apply until acknowledged
    Block,                // fail startup if drift is detected; too dangerous to ignore
}

pub trait Reconcilable: Additive {
    /// Defaults to `Block` — the safest behavior. Implementors that need
    /// `AutoApply` or `RequireConfirmation` override this explicitly.
    const DRIFT_ACTION: ReconcileAction = ReconcileAction::Block;

    /// Returns Some(description) if declared differs from stored, None if in sync.
    fn detect_drift(declared: &Self::Item, stored: &Self::Item) -> Option<String>;
}
```

> When startup detects drift on a `RequireConfirmation` type, it emits a structured WARN log event before writing to `config_drift` and continues using the **stored** value — behavior is unchanged until the operator acts:

Reviewer: I think some more detail on "drift" is needed.
-  what does the system do while waiting for confirmation?
- how does an operator fix a "blocked" change if the system won't start?
- once a change is confirmed, does it need to be re-confirmed?
- how does the operator know that confirmations are pending?  do they need to use the cli every time the systems are restarted?
- will the nico-cli handle the confirmations for all systems? (core-api, rest-api, etc)

Author: I am going to drop the `drift` concept from this initial design.  To your point, it needs more thought

# ConfigStore Design

> **Date:** 2026-06-25

Reviewer 2: A textual date (June 25, 2026) removes ambiguity, since Americans do goofy things like MM-DD-YYYY. Or, specify the format right in the text, like "Date (YYYY-MM-DD): 2026-06-25"

## Problem

Configuration in the `nico` codebase is consumed inconsistently across several dimensions:

- Some values can be updated at runtime via API; others require a service restart, with no convention distinguishing them.
- API shapes for creating and removing objects vary across resource types.
- Some resources have two configuration mechanisms (config file and API), creating confusion about which is authoritative.
- Some objects can be safely removed; others cannot because other resources depend on them — but this is not structurally enforced anywhere.
- Config is read via a mix of direct Figment calls, `std::env::var`, and bespoke patterns — the lookup shape differs per crate.
- External services (e.g., the Go-based `rest-api`) must independently configure values that api-core already owns, leading to duplication and drift.

The goal is a single, typed read interface that all components use to access configuration, a  single creation path
for additive objects regardless of whether they originate from a config file or an API call, and enforcement
of which objects can be safely removed.

## Scope

> 
> **In scope:**
> - `Configurable` trait — typed read interface for service config

Reviewer 2: Blank lines around lists
```suggestion
**In scope:**

- `Configurable` trait — typed read interface for service config
```

- `Additive` trait — objects seeded from config and addable via API (route_servers, network_segments, resource_pools)
- `Removable` trait — additive objects that can be safely removed (no dependents — like route_servers)
- `FileConfigStore` — Figment-backed implementation (TOML + overlays + env vars)
- `GrpcConfigStore` — read-only Rust implementation backed by api-core's gRPC ConfigService
- gRPC `ConfigService` in api-core serving config to external clients (Go and Rust)
- `config-store` crate housing the traits, errors, and implementations
- Unified creation path for additive objects (same service-layer logic for file seeding and API calls)
- Service-layer `object_service::create_or_skip<T: Additive>` and `object_service::remove<T: Removable>` as the type-enforced mutation entry points

> 
> **Out of scope (deferred):**
> - Write/update paths for service config — changes require a config file edit and restart

Reviewer 2: ```suggestion
**Out of scope (deferred):**

- Write/update paths for service config — changes require a config file edit and restart
```

> - File watching / hot reload

Reviewer 3: Wouldn't file watching for TLS certs be crucial for the config store to ensure rotated certs are immediately available?

Author: This refers to hotloading configuration changes.  So if you changed the location of the TLS cert in the config, the api pod would need to be restart

## Two Categories of Configuration

### Category 1: Service Configuration

Behavioral parameters set at deployment time. Never modified via API. Changing them requires editing the config file and restarting the service. Examples: TLS certificate paths, listen address, database URL, DHCP servers, ASN, controller timing parameters.

- Source of truth: config file (TOML), with optional site overlay and env var overrides.
- Read by api-core at startup via `FileConfigStore`.
- Read by external services (e.g. `rest-api`) over gRPC from api-core after those services start.

### Category 2: Additive Objects

Entities with a logical "this object should exist" definition expressible in a config file and also creatable via API. They accumulate over time; deletion is rare at best (route servers), impossible at worst due to dependents (VPCs, NetworkSegments, ResourcePools).

- Config file and API are **symmetric**: both express "this object should exist" and both go through the same service-layer code path.
- Source provenance (config vs. API) is retained for display and audit only, not to change behavior.

## Config File Structure Convention

> All configuration sections must be named TOML tables. Top-level flat keys (currently scattered at the document root in `carbide-apiconfig.toml`) move under a `[general]` section. Both the base config and any site overlay must use `[general]` for these fields.

Reviewer 3: Should we consider defining a JSON or toml schema to give these definitions more stability? [Taplo](https://github.com/tamasfe/taplo) seems to offer schema management.

Author: I was just thinking about that :)  Ill take a look at Taplo and get something added

Author: I added an example schema for the `general` section of the config file.  The JSON schema file allows us to use CLI tools like `Taplo` as well as use libraries like the `jsonschema`  to validate configuration data within nico-api.

> 
> Before:
> ```toml

Reviewer 2: Also need blanks around fenced code blocks
````suggestion
Before:

```toml
````

listen = "[::]:1081"
database_url = "postgres://a:b@postgresql"
asn = 123

[tls]
identity_pemfile_path = "/path/to/cert"
```

> 
> After:
> ```toml

Reviewer 2: ````suggestion
After:

```toml
````

[general]
listen = "[::]:1081"
database_url = "postgres://a:b@postgresql"
asn = 123

[tls]
identity_pemfile_path = "/path/to/cert"
```

Site overlay files follow the same structure and are merged by Figment — site keys win, base keys not present in the site overlay are preserved.

## Trait Hierarchy

```mermaid
graph TD
    Configurable --> Additive
    Additive --> Removable
```

> ### `Configurable`

Reviewer 2: In general, avoid formatting in headings. It's clear here that these H3s are the traits from the diagram.
```suggestion
### Configurable
```

Implemented by any type representing a config section. `KEY` is the dotted TOML path (e.g. `"tls"`, `"general"`, `"auth.cli_certs"`). Field-level defaults are handled by serde's `#[serde(default)]` — the store is not involved.

```rust
pub trait Configurable: DeserializeOwned {
    const KEY: &'static str;
}
```

Types implementing only `Configurable` are **file-only** — the only way to change them is to edit the config file and restart.

> ### `Additive`

Reviewer 2: ```suggestion
### Additive
```

```rust
async fn list<T: Additive>(&self) -> Result<Vec<T::Item>, ConfigError> {
    self.get::<T>().await.map(T::items)
}
```

Adding is always safe and has no preconditions. The config file and API both go through the same service-layer creation path at startup and at API call time respectively. Today this is implemented as bespoke per-resource functions (`create_initial_vpcs`, `reconcile_pool_defs`, `replace` for route servers, etc.). Part of the implementation work is converging these toward a consistent idempotent pattern.

Types implementing `Additive` but **not** `Removable` cannot be safely deleted — other resources depend on them.

The startup behavior differs by removability:
- **Removable** types: startup replaces the config-file-sourced entries in the database with whatever the current config declares. This is the "desired state = config file" model (route servers today).
- **Non-removable** types: startup runs `create_or_skip` — new entries are created; existing entries are left untouched. When an incoming item differs from the stored row, `create_or_skip` emits a `tracing::warn!` identifying the key and the mismatch (drift logging). The stored row is not modified.

> ### `Removable`

Reviewer 2: ```suggestion
### Removable
```

Extends `Additive`. Implemented by types whose entries can be safely removed with no side effects — they have no dependents. Removal is as routine as addition.

```rust
pub trait Removable: Additive {}
```

The type system enforces this: `object_service::remove::<T>()` does not compile unless `T: Removable`. There is no runtime guard needed.

Example: `RouteServersConfig` implements `Removable` because removing a route server has no effect on other resources.

> 
> ## Resource Classification

Reviewer 2: Delete extra blank lines. Yes the linter is picky :)
```suggestion
## Resource Classification
```

| Resource | Traits | Startup behavior | Reasoning |

> |---|---|---|---|

Reviewer 2: Making the header rows easier to read (easier to keep track of the alignment row this way, too)
```suggestion
| Resource | Traits | Startup behavior | Reasoning |
|----------|--------|------------------|-----------|
```

| `TlsConfig` | `Configurable` | N/A | File-only; restart to change |
| `GeneralConfig` | `Configurable` | N/A | File-only; restart to change |
| `SiteExplorerConfig` | `Configurable` | N/A | File-only; restart to change |
| `MachineStateControllerConfig` | `Configurable` | N/A | File-only; restart to change |
| `ResourcePoolsConfig` | `Additive` | `create_or_skip`; drift logged as `warn` | Removal unsafe; machines depend on pool addresses |
| `NetworksConfig` | `Additive` | `create_or_skip`; drift logged as `warn` | Removal unsafe; change is dangerous |
| `VpcDefinitions` | `Additive` | `create_or_skip`; drift logged as `warn` | Removal unsafe; other resources are scoped to a VPC |
| `NetworkSegmentsConfig` | `Additive` | `create_or_skip`; drift logged as `warn` | Removal unsafe; machines are assigned to segments |

> | `RouteServersConfig` | `Additive + Removable` | Replace (desired state = config file) | No dependents; add/remove/replace freely |

Reviewer 2: Any reason not to alpha-sort this?

```suggestion
| `GeneralConfig` | `Configurable` | N/A | File-only; restart to change |
| `MachineStateControllerConfig` | `Configurable` | N/A | File-only; restart to change |
| `NetworksConfig` | `Additive` | `create_or_skip`; drift logged as `warn` | Removal unsafe; change is dangerous |
| `NetworkSegmentsConfig` | `Additive` | `create_or_skip`; drift logged as `warn` | Removal unsafe; machines are assigned to segments |
| `ResourcePoolsConfig` | `Additive` | `create_or_skip`; drift logged as `warn` | Removal unsafe; machines depend on pool addresses |
| `RouteServersConfig` | `Additive + Removable` | Replace (desired state = config file) | No dependents; add/remove/replace freely |
| `SiteExplorerConfig` | `Configurable` | N/A | File-only; restart to change |
| `TlsConfig` | `Configurable` | N/A | File-only; restart to change |
| `VpcDefinitions` | `Additive` | `create_or_skip`; drift logged as `warn` | Removal unsafe; other resources are scoped to a VPC |
```

> ## `ConfigStore`

Reviewer 2: ```suggestion
## ConfigStore
```

    // failures are never silently swallowed.
    pub async fn get_or_default<T: Configurable + Default>(&self) -> Result<T, ConfigError> {
        match self.get::<T>().await {
            Ok(v) => Ok(v),
            Err(ConfigError::NotFound { .. }) => Ok(T::default()),
            Err(e) => {
                tracing::warn!(key = T::KEY, "config fetch failed; using default only for NotFound");
                Err(e)
            }
        }
    }

    // Category 2: read the current set of additive objects.
    pub async fn list<T: Additive>(&self) -> Result<Vec<T::Item>, ConfigError> {
        self.get::<T>().await.map(T::items)
    }
}
```

Each store type implements `ConfigBackend` and exposes a constructor that produces a `ConfigStore`. Callers never hold the inner store type after construction.

`ConfigStore::get` is intended for startup reads and infrequent operational fetches. `#[async_trait]` desugars `get_raw` to a boxed future, adding one heap allocation per call. Services that need config values on every request should call `store.get()` once at startup, cache the result, and read the cached value per-request — not call `store.get()` in the hot path.

> ## `ConfigError`

Reviewer 2: ```suggestion
## ConfigError
```

`ConfigError` messages must not expose raw config values. Error messages include only structural metadata: the section key name (`key: &'static str`) and error category. They never embed field values such as database passwords, TLS key material, or token strings.

`ConfigError::Deserialize` carries only the section key — no source error and no raw value — because both `figment::Error` and `serde_json::Error` can include the value that failed to parse in their `Display` output. Dropping the source at the error boundary prevents credentials from leaking into logs. A unit test shall assert that constructing a `ConfigStore` from TOML containing a `database_url` with a password does not produce that password in any `ConfigError` display string.

`ConfigError::Io` includes `path` in its display, which is acceptable for ordinary config file paths (e.g. `/etc/carbide/config.toml`). If the future `figment_file_provider_adapter` feature is added and secret values are read from files under `/run/secrets/`, `Io` errors from those reads should redact the path to its directory component only (e.g. `/run/secrets/<redacted>`) to avoid exposing which secret is being read.

> ## `FileConfigStore`

Reviewer 2: ```suggestion
## FileConfigStore
```

### Builder

`FileConfigStoreBuilder` is the builder type. All three builder methods (`builder`, `with_overlay`, `with_env_prefix`) are infallible — they record paths and settings without doing I/O. All file reads happen in `build()`, which is the single `Result`-returning boundary.

```rust
impl FileConfigStore {
    pub fn builder(base: impl AsRef<Path>) -> FileConfigStoreBuilder;
}

impl FileConfigStoreBuilder {
    /// Layer a site-specific TOML overlay. Later overlays take precedence.
    pub fn with_overlay(self, path: impl AsRef<Path>) -> Self;

    /// Layer environment variable overrides with the given prefix.
    /// Uses `__` as a section separator (e.g. `CARBIDE_API_TLS__CERT_PATH`).
    pub fn with_env_prefix(self, prefix: &str) -> Self;

    /// Reads all registered files, runs the Figment merge, stores the result
    /// in-memory, and wraps it in a ConfigStore. All file I/O happens here —
    /// missing or malformed files surface as ConfigError::Io.
    pub fn build(self) -> Result<ConfigStore, ConfigError>;
}
```

> ### `ConfigBackend` implementation

Reviewer 2: ```suggestion
### ConfigBackend implementation
```

        get_nested(&self.data, key)
            .cloned()
            .ok_or(ConfigError::NotFound { key })
    }
}
```

### Test constructor

```rust
impl FileConfigStore {
    /// Accepts an inline TOML string instead of a file path. Performs the same
    /// Figment merge and JSON snapshot as build(). Does not support overlays or
    /// env var layers.
    #[cfg(any(test, feature = "test-utils"))]
    pub fn from_toml_str(s: &str) -> Result<ConfigStore, ConfigError>;

    /// Merges two inline TOML strings in the same order as
    /// builder().with_overlay(): base values first, then overlay on top.
    /// The overlay wins on any key present in both.
    #[cfg(any(test, feature = "test-utils"))]
    pub fn from_two_toml_strs(base: &str, overlay: &str) -> Result<ConfigStore, ConfigError>;
}
```

### Call site

```rust
let store: ConfigStore = FileConfigStore::builder("/etc/carbide/config.toml")
    .with_overlay("/etc/carbide/site.toml")
    .with_env_prefix("CARBIDE_API_")
    .build()?;

let tls: TlsConfig = store.get().await?;
let explorer: SiteExplorerConfig = store.get_or_default().await?;
let pools: Vec<PoolConfig> = store.list::<ResourcePoolsConfig>().await?;
```

> ## gRPC `ConfigService`

Reviewer 2: ```suggestion
## gRPC ConfigService
```

api-core exposes a `ConfigService`. External clients use it to fetch config values that cannot be read from a local file.

### When external services fetch config

External services (e.g. the Go `rest-api`) have two tiers:

1. **Startup config** — values required before the service can connect to anything (database, Temporal, TLS, listen port). Read from the service's own local config file at process start.

> 2. **Operational config** — values that can be fetched after the service is running (JWT issuers, auth parameters, rate limiter settings, site-level parameters). Fetched from api-core's `ConfigService` over gRPC.

Reviewer 2: We can omit the numbering here

```suggestion
**Startup config** — values required before the service can connect to anything (database, Temporal, TLS, listen port). Read from the service's own local config file at process start.

**Operational config** — values that can be fetched after the service is running (JWT issuers, auth parameters, rate limiter settings, site-level parameters). Fetched from api-core's `ConfigService` over gRPC.
```

```rust
// In api-core alongside the ConfigService implementation.
// This is the authoritative, reviewable list of what GetRawSection exposes.
const REMOTE_SECTION_ALLOWLIST: &[&str] = &[
    IssuersConfig::KEY,       // "issuers"
    AuthConfig::KEY,          // "auth"
    RateLimiterConfig::KEY,   // "rate_limiter"
];
```

**Schema compatibility.** `GetRawSection` returns JSON with no schema version tag. If api-core renames or restructures a config section, any deployed `GrpcConfigStore` client targeting that section will receive a successful RPC response that fails deserialization as `ConfigError::Deserialize`. There is no backward-compatibility guarantee on the JSON shape emitted by `GetRawSection`. `GrpcConfigStore` clients must be updated and redeployed alongside api-core when the config schema changes. This is acceptable when api-core and its Rust service consumers share the same deployment lifecycle.

api-core serves both surfaces from its in-memory config store. Request messages for typed RPCs are empty (or carry a version field for future cache validation).

> ## `GrpcConfigStore`

Reviewer 2: ```suggestion
## GrpcConfigStore
```

A `ConfigStore` backed by api-core's `ConfigService` gRPC endpoint. Intended for future Rust services deployed separately from api-core that need operational config at runtime without a local config file.

### Constructor

```rust
impl GrpcConfigStore {
    pub async fn connect(endpoint: Uri) -> Result<ConfigStore, ConfigError>;
}
```

> ### `ConfigBackend` implementation

Reviewer 2: ```suggestion
### ConfigBackend implementation
```

The practical consequence is that API-added entries (e.g. a route server added via `nico-admin-cli`) are ephemeral: they survive until the next api-core restart, after which only the entries declared in the config file remain. Operators who want an API-added entry to persist across restarts must add it to the config file. This keeps the config file as the single authoritative source for `Removable` types and avoids a class of "where did this entry come from?" operational confusion.

The transaction serializes against concurrent `object_service::remove` and `object_service::create_or_skip` calls on the same type.

## Crate Structure

The crate targets the Rust 1.90.0 stable toolchain (pinned in `rust-toolchain.toml`) on Linux x86-64 and Linux aarch64; no nightly features are used.

> ```

Reviewer 2: Code blocks all need a language. Text is a safe default for all renderers.
````suggestion
```text
````

Author: None of this is published as docs, so I don't think it really matters

Author: Fixed

**Additive object seeding tests** verify idempotency: running startup seeding twice produces the same DB state as running it once.

## Implementation Sequencing

Implement in this order to keep each phase independently testable:

1. **`FileConfigStore` + `get` / `list` + `get_nested` + unit tests** — the file store is entirely self-contained and is the only dependency of the seeding path. All unit tests described in the Testing Strategy section can be validated at this stage.
2. **`object_service::create_or_skip` + startup seeding refactor** — depends only on `FileConfigStore` and the existing DB layer. The full config-file → `store.list` → `create_or_skip` → DB flow can be exercised against a real store without any gRPC scaffolding.
3. **`object_service::remove` + `Removable` replace transaction** — builds on the same DB layer; no gRPC dependency.

> 4. **gRPC `ConfigService` + `GrpcConfigStore`** — add once a consuming service actually needs remote config. At that point `FileConfigStore` is stable, the contract is exercised, and the gRPC layer has a clear, tested interface to proxy.

Reviewer 2: Unless a file is meant to be read as plain text, numbering everything as 1 renders correctly in any preview, and is easier to maintain. Not a big issue in a 4-item list, but you'd be surprised how often these wind up being something like 1-2-3-5 as a doc gets worked on.
```suggestion
1. **`FileConfigStore` + `get` / `list` + `get_nested` + unit tests** — the file store is entirely self-contained and is the only dependency of the seeding path. All unit tests described in the Testing Strategy section can be validated at this stage.
1. **`object_service::create_or_skip` + startup seeding refactor** — depends only on `FileConfigStore` and the existing DB layer. The full config-file → `store.list` → `create_or_skip` → DB flow can be exercised against a real store without any gRPC scaffolding.
1. **`object_service::remove` + `Removable` replace transaction** — builds on the same DB layer; no gRPC dependency.
1. **gRPC `ConfigService` + `GrpcConfigStore`** — add once a consuming service actually needs remote config. At that point `FileConfigStore` is stable, the contract is exercised, and the gRPC layer has a clear, tested interface to proxy.
```

Author: Fixed
